src/core/pipe/pipe.middleware.js

```diff
@@ -0,0 +1,37 @@
+export default class Middleware {
+
+	constructor(stages) {
+		this.stages = stages || [];
+	}
+
+	run(ctx, data = []) {
+		return pipeline(
+			this.stages.map(s => data => new Promise(resolve => {
+				s(data, ctx, resolve);
+			})),
+			data);
+	}
+}
+
+function pipeline(stages, initialData) {
+	let index = 0;
+
+	return new Promise((resolve, reject) => {
+		invoke(initialData);
+
+		function invoke(data) {
+			let stage = stages[index++];
```

Comment:
can be that stages will mutate during invokation? maybe we need to make copy of it?